enclave/core/sgx/once.c

```diff
@@ -5,34 +5,74 @@
+#define FUNC_INVOKED 2
+
 oe_result_t oe_once(oe_once_t* once, void (*func)(void))
 {
     if (!once)
         return OE_INVALID_PARAMETER;
 
+    oe_once_t status = *once;
     /* Double checked locking (DCLP). */
-    oe_once_t o = *once;
-
     /* DCLP Acquire barrier. */
     OE_ATOMIC_MEMORY_BARRIER_ACQUIRE();
-    if (o == 0)
-    {
-        static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
 
-        oe_spin_lock(&_lock);
-
-        if (*once == 0)
+    /*
+      Use an atomic-acquire load operation to check if the function has not been
+      invoked. If the function has already been invoked, there is nothing to do.
+      If the function is being invoked, then this thread must wait for the
+      function invocation to complete. Otherwise, this thread can try to take
+      ownership of invoking the function
+    */
+    if (status != FUNC_INVOKED)
+    {
+        /*
+          Multiple threads could reach here simultaneously after checking
+          whether the function has been invoked or not. Only one of them must
+          now invoke the function and others must wait for the function
+          invocation to complete.
+          To determine who gets to invoke the function, each thread atomically
+          compares the value of once to FUNC_NOT_INVOKED and if equal, sets the
+          value to FUNC_BEING_INVOKED to signal to other threads that the
+          function is being invoked. The return value of
+          __sync_val_compare_and_swap determines which thread takes ownership.
+          If __sync_val_compare_and_swap returns FUNC_NOT_INVOKED, then that
+          means this thread successfully set once to FUNC_BEING_INVOKED, and can
+          now safely call func. If __sync_val_compare_and_swap returns
+          FUNC_BEING_INVOKED, this means another thread's
+          __sync_val_compare_and_swap succeeded first and therefore this thread
+          now has to wait for the other thread to complete (ie wait for once to
+          become FUNC_INVOKED. If __sync_val_compare_and_swap returns
+          FUNC_INVOKED, that means another thread has already called the
+          function and marked the once as complete. This thread can safely
+          proceed.
+        */
+        oe_once_t retval = __sync_val_compare_and_swap(
+            once, FUNC_NOT_INVOKED, FUNC_BEING_INVOKED);
+        if (retval == FUNC_NOT_INVOKED)
         {
             if (func)
                 func();
 
-            /* DCLP Release barrier. */
             OE_ATOMIC_MEMORY_BARRIER_RELEASE();
-            *once = 1;
+            *once = FUNC_INVOKED;
+        }
+        else if (retval == FUNC_BEING_INVOKED)
+        {
+            /*
+              Another thread is invoking the function. Wait for that thread to
+              finish the invocation and mark the once variable to FUNC_INVOKED.
+            */
+            while (__sync_val_compare_and_swap(
+                       once, FUNC_BEING_INVOKED, FUNC_BEING_INVOKED) !=
```

Comment:
Based of the comment, shouldn't this be:
`while (!__sync_bool_compare_and_swap(once, FUNC_INVOKED, FUNC_INVOKED))`? (i.e CAS loop until it *once == FUNC_INVOKED).

It seems that the current logic would fail in the following condition:
```C
Thread1: else if (retval == FUNC_BEING_INVOKED)
Thread2: *once = FUNC_INVOKED
Thread1: while (__sync_val_compare_and_swap(once, FUNC_BEING_INVOKED, FUNC_BEING_INVOKED) != FUNC_BEING_INVOKED) // infinite loop since it will always return FUNC_INVOKED.
```

Reply:
@gupta-ak

Yes you are right.
 The intent was that if a thread sees retval == FUNC_BEING_INVOKED, then it needs to wait for the invocation to complete. This could be expressed as:

- Wait until *once == FUNC_INVOKED
- Or wait while *once == FUNC_BEING_INVOKED

I don't remember which one we were trying to code. The first one is more easier to read.
So we should use the loop you provided.

Is there a better way to express _loop until *once == FUNC_INVOKED_ ? 
Can we somehow hide the compare and swap and expresses it as an _atomic compare_ or _atomic load and then compare_?

Reply:
I think this works:
```
while (*((volatile oe_once_t*) once) != FUNC_INVOKED) {}
ACQUIRE_BARRIER
```

Just we have to make sure that the for loop doesn't get optimized and that there's an acquire barrier to synchronize with the release barrier.